### ISSUE TITLE
Processing Improvement Discussion (ImprovedProcessinf branch)

### ISSUE THREAD

**Issue by ilia3101** [ilia3101/MLV-App#158]
Originally this issue was me asking a question about compiling speed, but it has become a discussion about implementing many features in processing. So some of the comments at the start are irrelevant.

Original first post...

---------------------------------------------------------------------------------------

Never ever had it compile without it, however I do end up with a 'mlvapp' execultable that I run.

Don't know what this means:

```
tar -C /home/ilia/bin -xvJf /home/ilia/MLV-App/platform/qt/FFmpeg/ffmpegLinux.tar.xz --strip=1 ffmpeg-4.0-64bit-static/ffmpeg
tar: /home/ilia/bin: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
make: *** [Makefile:481: mlvapp] Error 2
```

**Comment by ilia3101** [comment]
Also is there any chance of making it compile faster? Takes so long even with -j4 on my 'new' i5 5200u dell :cry:

**Comment by masc4ii** [comment]
Puh... no idea. Have you tried checking out - compiling from the start? I guess yes.

Yes, we could make it faster. For me the main problem for slow compilation is MainWindow.cpp. This file is soooo huge. But I don't know how to split it in small packages. There are so many functions which need access here and there. 😞

**Comment by ilia3101** [comment]
Ah does ffmpeg need to be extracted first?

Don't make faster compiling a priority don't worry for now.

**Comment by ilia3101** [comment]
Can we do a quick release in approximately 2 days

**Comment by ilia3101** [comment]
@masc4ii do you know why it might be crashing every time I press the open mlv or open session? Even without my changes... no file picker window shows.

**Comment by masc4ii** [comment]
The command in the .pro file should extract ffmpeg automatically, if I remember right.

We can realease whenever you like. Do you plan something special?

Those crashes I had a long time ago. But I don't remember the reason... sounds like a problem with the libraries. Under (left bar) Project, (chose project) Run, there are some checkboxes with DYLD_LIBRARY_PATH... maybe try playing around there.

**Comment by ilia3101** [comment]
I would like to do a release on April 1 as it is an important day in magic lantern and I am going on holiday on April 4 so it seemed like a suitable day. Hopefully a couple of improvements I'm doing will be ready by tomorrow or on the 1st.

And I will try installing qt creator and doing what you say, just using terminal right now.

**Comment by masc4ii** [comment]
Oh... terminal?! OMG... never tried that 😄 
Let's see if I can get ready my things too...

**Comment by ilia3101** [comment]
I am in a new branch right now, BetterProcessing. What kind of things are you planning, processing?

I have been using terminal almost the whole time :rofl:

**Comment by ilia3101** [comment]
And it doesn't crash when I run from Qt creator arghhhhhhhhH!!!!!!!!!!!!!!!1 :(((

But when i do ./mlvapp it crashes again :(

**Comment by masc4ii** [comment]
Japp... as expected. Library problem. 

Yes, the noise generator is in code already, but there is no parameter yet to set it up. Same for the better blur for Highlights/Shadows. You can switch it in the code only atm. I use it already and I am very satisfied. Maybe it is also okay to just use the better blur... what do you think?
What does better processing? ACES?

**Comment by ilia3101** [comment]
I completely agree your better blur looked incredible for highlights and shadows, just switch it out instead of my blur already.

The better processing branch is what I've been wanting to do, but not just ACES.

**Comment by ilia3101** [comment]
> Japp... as expected. Library problem.

Do you know how to recreate the way Qt runs but it in terminal? Is it some kind of environment variable that needs to be set? As I don't like switching between two text editors all the time to run it.

**Comment by masc4ii** [comment]
It might happen if you have different Qt versions installed on your system. If you have just one Qt, you'll (hopefully) never have this problem - at least I know this only from systems where some Qt library was installed systemwide, but you program with a different version.

**Comment by ilia3101** [comment]
Don't think I have different Qt installed this is confusing. I'll figure it our later, for now I'll just drag files on for testing.

**Comment by ilia3101** [comment]
@masc4ii Would you be able to add a profile section to the interface? It would need to have a profile selection menu at the top, then some menus and options below it. The lower options would be stuff like output gamut, tonemapping function, gamma etc, and would all change at once when a profile is selected, and when the user changes any of them, the profile will become 'custom'  - is this possible? or too much?

Compatibility with old files will be kept as all the same profiles will be available still.

**Comment by masc4ii** [comment]
Grain-Gen is working now. Yay.

If you like, we could do try that. So the profile (and also Cam-Matrix?) will get a separate section somewhere? And old profile will be splitted to Profile and ...? What should be the options in which combobox?
Not sure if we can have compatiblity... this sounds difficult to realize. But we could try also this, if you have an idea how.

**Comment by ilia3101** [comment]
1 min almost done mockup

**Comment by ilia3101** [comment]
Would look like this:

![Screenshot from 2019-03-30 21-40-25](https://user-images.githubusercontent.com/23642861/55282159-8f5bac80-5336-11e9-9e9d-09a7331eb102.png)

The items below the 'profile' would be set automatically when setting profile, there will be some getter functions on processingObject for finding their values after setting profile.

**Comment by masc4ii** [comment]
Whohow... this looks like a lot of work you plan on your side. The GUI should be doable... not sure if in one day, but doable.

**Comment by ilia3101** [comment]
It is a lot of work but not as much as it seems, I have come up with the easiest possible way to do it :wink: (keeping much of the old stuff)

updated mockup:
<img src="https://user-images.githubusercontent.com/23642861/55282252-7653fb00-5338-11e9-8f89-6be0a105a4ab.png" width="200">

**Comment by ilia3101** [comment]
https://github.com/ilia3101/MLV-App/commit/52cb5f9e302f44e4015ab843e8a7eb2268e51568

**Comment by ilia3101** [comment]
https://github.com/ilia3101/MLV-App/commit/52cb5f9e302f44e4015ab843e8a7eb2268e51568

@masc4ii @dannephoto @bouncyball-git 

Can anyone test if the BetterProcessing branch crashes on windows/mac or anything like that.

And no, it is not yet any use for actual processing so don't use it.

**Comment by seescho** [comment]
> Never ever had it compile without it, however I do end up with a 'mlvapp' execultable that I run.
> 
> Don't know what this means:
> 
> ```
> tar -C /home/ilia/bin -xvJf /home/ilia/MLV-App/platform/qt/FFmpeg/ffmpegLinux.tar.xz --strip=1 ffmpeg-4.0-64bit-static/ffmpeg
> tar: /home/ilia/bin: Cannot open: No such file or directory
> tar: Error is not recoverable: exiting now
> make: *** [Makefile:481: mlvapp] Error 2
> ```

Do you have "bin"-directiory in your home-dir (Some distributions come without one, I heard)? If no, create one:
`mkdir ~/bin`
Edgar

**Comment by masc4ii** [comment]
GUI elements painted and commited. Next we need all elements inside the comboBoxes. With QtCreator just doubleclick those items and write the texts in the list. Then I'll create the callbacks and we can shot on your macro's/functions.

**Comment by masc4ii** [comment]
I added the most important callbacks for you. Some of them already have the link to processing unit. Anyway, the order of the comboBox entries has to be the same as your defines (if you create them for your functions). 
Expected: no receipt support at all atm. This means, switching between clips will lead to undefined behavior. I think you should change all combo's after switching a clip to get a defined look.

**Comment by ilia3101** [comment]
I'll push more code in 1-2 hours when I'm home, it's working but image looks like rawtherapee right now, hopefully can fix that.

**Comment by ilia3101** [comment]
The log functions should be in tonemapping menu

**Comment by ilia3101** [comment]
Also processing gamut should have all the same options as output gamut

**Comment by ilia3101** [comment]
Wait why is Linear an option in gamut? Is it accidentally left over?

**Comment by ilia3101** [comment]
> The log functions should be in tonemapping menu

I will add some macro values and get set function

**Comment by ilia3101** [comment]
Creative adjustments is really broken, not allowed at all in tonemapped/standard/film mode

**Comment by ilia3101** [comment]
I have a feeling something is really wrong with rec709 it looks so baaaaaaaaaaaaaaaaaaaad, adobeRGB even looks more like old mlv app's rec709 output. WTDFGfkjdsfghksdhfkajdsdfha

**Comment by ilia3101** [comment]
Well even using AdobeRGB, we can see improvement...

OLD, see clipping on green thing above orange thing:
![Screenshot from 2019-03-31 19-46-48](https://user-images.githubusercontent.com/23642861/55293378-df497a80-53ed-11e9-8262-4a3ab52c561f.png)

NEW- no clipping
![Screenshot from 2019-03-31 19-47-01](https://user-images.githubusercontent.com/23642861/55293389-fbe5b280-53ed-11e9-8f7f-460fd836b24a.png)

New looks less saturated as it's AdobeRGB

**Comment by masc4ii** [comment]
I thought you forgot linear. @seescho will still need it... 
Okay, so I copy all entries from output gamut to processing gamut. 

>Creative adjustments is really broken, not allowed at all in tonemapped/standard/film mode - not sure why. Is it in the interface?

If it is greyed out sometimes... yes. If it affects the image, it is in the C backend.

I wrote many lines for receipt support... will commit soon.

**Comment by ilia3101** [comment]
Well it won't be in gamut, it will be back as gamma set at 1.0 and no tonemapping function

**Comment by masc4ii** [comment]
Okay. 
Do we still need "Allow creative adjustments" if we add profile "Custom" ? Isn't this the same?

**Comment by ilia3101** [comment]
Allow creative adjustments should still be there I think, and if it is disabled, that's when it becomes "Custom" (showing that it's no longer real log)

**Comment by masc4ii** [comment]
Ah yes... forget about... already saw too much code today... LOL

**Comment by masc4ii** [comment]
Whoooaaa... the rules for all those comboBoxes will be hard to implement. I get crazy.

**Comment by ilia3101** [comment]
The way it should work is: The "Processing profile" combobox will call the profile setting function - after this you should call getter functions to set the values of the combo boxes (and gamma slider etc) below it, no special rules in Qt code.

Only rule I think should be to change profile to 'custom' when any of the other combo boxes or settings are changed by the user

**Comment by ilia3101** [comment]
Or is there something I'm missing that makes it really difficult? I am really sorry if adding MASXML support will be difficult

**Comment by masc4ii** [comment]
Not sure yet how it works best without making the interactive rulez between all those boxes to difficult.

**Comment by ilia3101** [comment]
Setting the "Processing profile" setting will affect the states of the settings below (inside of processing), so I think there should be a updater function called after setting a profile that reads the settings from the processing object and updates them in the interface.

**Comment by ilia3101** [comment]
I will add a processingSetTonemappingFunction(processing, int function) as well as processingGetTonemappingFunction, as well as defines of the functions so the tonemapping combo box can function

**Comment by masc4ii** [comment]
Okay. I'll try it this way. I expect some fun with the receipts in general... not only MASXML (this should work out of the box if receipts are working).

**Comment by ilia3101** [comment]
I apologize for using the word function 100000 times in that comment I hope it gets the message across of what I will do

**Comment by masc4ii** [comment]
I get crazy... if Profiles callback sets the other boxes via getters, their callback functions set Profiles box back to custom... 😢

**Comment by masc4ii** [comment]
Okay... a part is working now. Looks not too bad. But all clips imported look so flat until I change profile to something else and switch back. Don't understand what happens. Sry.

**Comment by ilia3101** [comment]
Sorry for my little absense.

> Okay... a part is working now. Looks not too bad. But all clips imported look so flat until I change profile to something else and switch back. Don't understand what happens. Sry.

Yeah strange that this is happening.

**Comment by ilia3101** [comment]
Sorry I'm really tired. I'll have more tomorrow. April 1st will definitely be missed for having it ready. Also needs some work to get the colours right, as reds look really bad in rec709 and there is huge amounts of noise everywhere.

**Comment by dannephoto** [comment]
Holymoly. I missed the color party completely. Nice progress!

**Comment by masc4ii** [comment]
@ilia3101 : I found the reason for the flat pictures: it has to do with "Allow Creative Adjustments". If I never grey out the checkbox, I just have to enable it, and the pictures aren't flat anymore. In a commit you delete "use_rgb_curves"... this enabled it before - this isn't there anymore, just allow_creative_adjustment is still there. So e.g. the curve only works, if the checkbox is set.

I have a question: when do we grey out Allow Creative Adjustment, and when we don't? Especially in "Custom" it makes not always sense to have it enabled...

**Comment by ilia3101** [comment]
Ah yes, sounds like a correct explanation. I want to have all creative adjustments enabled/disabled just by one variable (allow_creative_adjustment), including curves. I'm sure it's just a matter of changing a couple of if statements, but I'm not sure where to do it. Do you think you could fix this easily?

**Comment by ilia3101** [comment]
I see you have added special tonemapping functions for very accurate rec709 and sRGB gammas, is it ok if I don't include them for now and just do it as a simple output gamma value for those profiles?

I promise to bring those accurate functions back later. This is just to make it a little simpler for now.

**Comment by ilia3101** [comment]
> I have a question: when do we grey out Allow Creative Adjustment, and when we don't? Especially in "Custom" it makes not always sense to have it enabled...

Ah good question, I think we should just not grey it out anymore. So now it could be useful to compare how the image looks before and after.

**Comment by ilia3101** [comment]
@masc4ii Is it physically possible to transfer your rbf denoising to work inside apply_processing_object? Right before the final colour space conversion is done. I think it would create a better colour denoising result if it's done in the processing gamut.

**Comment by masc4ii** [comment]
Okay, thanks for the info. I can try to do this. I just needed the info how to do.

rec709 and srgb doesn't matter if it is not there. What I (and I think also others) need, is BMD-Film profile support, because there are many LUTs out there bringing a very nice looking picture. I would really miss it. 

All the RBF functions have to be rendered on full picture, because the enemy's name is "boarder". So you'll get very ugly stripes and artifacts when doing this in the threaded function. Maybe we need to build the same thread mechanism after denoising and grain, for post conversion? Or you use OMP there, this solution is mostly very very simple, because you just need one single line with `#pragma ...`

**Comment by ilia3101** [comment]
> What I (and I think also others) need, is BMD-Film profile support, because there are many LUTs out there bringing a very nice looking picture. I would really miss it.

Sure, it will stay.

> All the RBF functions have to be rendered on full picture, because the enemy's name is "boarder". So you'll get very ugly stripes and artifacts when doing this in the threaded function.

I see. Could split it in to two halves and apply it in between, but let's leave it for now.

I've been thinking of just using rec709 as the processing gamut (after doing tonemapping in a wider one to avoid the ugly clipping). Rec709 sounds limited, but if the lookup tables represented a range of values: something like -1.0 to 3.0, instead of the current 0.0-1.0, the gamut wouldn't be limited to rec709 at all, and would allow final output in wider gamuts. This may increase banding but it's still a really small risk.

I just think the rec709 gamut is very good for pleasing processing results and is why MLV App produced especially nice results before.

**Comment by bouncyball-git** [comment]
>  but if the lookup tables represented a range of values: something like -1.0 to 3.0, instead of the current 0.0-1.0, the gamut wouldn't be limited to rec709 at all, and would allow final output in wider gamuts.
> 
> I just think the rec709 gamut is very good for pleasing processing results and is why MLV App produced especially nice results before.

Hmmm... very interesting approach.

**Comment by masc4ii** [comment]
Could you please try out what we have now? GUI should work as expected now. 
Missing: Gamma and Tonemapping Function.
Strange: Processing Gamut is Rec.2020 if "Tonemapped" is selected. If I manually select Rec.709 it looks like in old versions. Is that correct?
@Ilia: have you seen the comment in the ML forum thread about ACES?

**Comment by ilia3101** [comment]
Thanks for the interface update, interesting to try out gamuts so easily. Will have more of the parameters setting functions soon.

**Comment by ilia3101** [comment]
@masc4ii Seems like there will have to be a "lighten" type parameter added to the processing profile section, it is the only way I can see to replicate the tonemapped profiles correctly, as the old way of "3.15 gamma" was very bad colour science grammar.

**Comment by ilia3101** [comment]
Can we add a mouse over explanation pop up to some of these parameters? I feel like it might end up being a confusing section.

**Comment by ilia3101** [comment]
> Strange: Processing Gamut is Rec.2020 if "Tonemapped" is selected. If I manually select Rec.709 it looks like in old versions. Is that correct? I would expect, Rec.709 is chosen automatically.
> @Ilia: have you seen the comment in the ML forum thread about ACES?

Very correct! I am still deciding what the default processing gamut should be, just put Rec2020 randomly for now. And setting it to rec709 should produce old results, so it is good that it does. I want to use something wider by default when we actually release, will work better for low light and things.

I know changing processing gamut right now causes HuGE changes in the output, uglier colour mostly, this will be fixed and should not be happening (only small changes should happen).

**Comment by bouncyball-git** [comment]
Tested with Danne's bluish clip with temp = 3000K:

Rec709/sRGB processing
![srgb](https://user-images.githubusercontent.com/17232233/55564365-ccde8200-5708-11e9-8fc1-6d080214c9b0.png)

Rec2020
![rec2020](https://user-images.githubusercontent.com/17232233/55564468-01ead480-5709-11e9-8778-0489af5a1d61.png)

aces ap0
![acesap0](https://user-images.githubusercontent.com/17232233/55564481-08794c00-5709-11e9-9354-6c52d1a73637.png)

ACES processed version looks less bluish. Multiplier function of WB code really needs to be revised right? @ilia3101

**Comment by dannephoto** [comment]
Yes, glad you brought this up. I also tested with another clip and same wb issue.
Really nice to see these color gamuts being implemented.
Now back to the wb issue. I am so frustrated to not knowing at all where to look for hiw to fix it. It kind of puzzles me how colors seems right but at the same time they crack into nlue in certain conditions.

**Comment by bouncyball-git** [comment]
Yeah color processing is very interesting bur IS another world. And needs so much dedication and time to comprehend it to usable degree to jump into Ilia's and Masc's processing development :). I'm off right now... no time at all.

Guys thank you for your time and goodies you are producing!

**Comment by masc4ii** [comment]
>Can we add a mouse over explanation pop up to some of these parameters? I feel like it might end up being a confusing section.

Sure. Tell me what phrase you would like on which element. Or, go into QtCreator, search elements in UI-Editor and write it. 😄 

I added the tonemap function now to the GUI. All LOG options seems to be very bright... but I think it is because gamma is not working yet.
If selecting a profile, "getTonemappingFunction" seems not to bring the correct value. It is always "None", but looks like "Reinhard".

**Comment by masc4ii** [comment]
Hi @ilia3101 , any progress on your side?

**Comment by ilia3101** [comment]
Hi sorry, gone on holiday. I was going to try spend some time on this, but I broke my 5D with a unlucky fall and not feeling much motication, not until I find another ML camera at least. Might ask on ML forum where I can find an exceptionally cheap 5D mark II.

@bouncyball-git thanks for doing that test, it really doesn't look great :(

@masc4ii or @dannephoto could any of you test that clip in Adobe software to see how that handles it?

**Comment by ilia3101** [comment]
Yeah, quite a few options are not working yet so I imagine log might not

**Comment by ilia3101** [comment]
And I know the results in this branch still don't look right, they reallly don't especially with red colours.

**Comment by masc4ii** [comment]
Oh noooo... not again! It broke in holiday? F***. Can understand you are not very motivated now. But it is a shame for this topic. I had hope, we solve it this time... Wish you much luck to find another 5D2 or 5D3. I think with other MLV camera's, you won't get happy. E.g. EOS M is a nice little toy with many great features, but quality from 5D2 is still another class.

To just have a matrice multiplication to get from one space into another is right, there is no offset needed? (for having the center (D65) always at the correct position) There are combinations between processing and output gamut, where all becomes red or blue... right.

I could compare in Resolve. I have no idea how to do in Lightroom (is it possible here?).

**Comment by bouncyball-git** [comment]
> I could compare in Resolve. I have no idea how to do in Lightroom (is it possible here?).

I think he meant develop just one DNG frame from that Dann's clip in adobe product.

**Comment by bouncyball-git** [comment]
> I broke my 5D with a unlucky fall and not feeling much motivation, not until I find another ML camera at least. Might ask on ML forum where I can find an exceptionally cheap 5D mark II.

Uhhh... not again, NOT AGAIN!!! 😁 (your sensor was not calibrated properly anyway). Get 5D3 and you'll be set up for cuople of years from now.

**Comment by ilia3101** [comment]
> Oh noooo... not again! It broke in holiday? F***. Can understand you are not very motivated now. But it is a shame for this topic. I had hope, we solve it this time... Wish you much luck to find another 5D2 or 5D3. I think with other MLV camera's, you won't get happy. E.g. EOS M is a nice little toy with many great features, but quality from 5D2 is still another class.

Useful to know that 5D2 is much better than EOS M, I was considering EOS M but I won't be getting it as main camera then. I will keep looking for a 5D2 or 3. There's something satisfying about using a 10 year old camera, 5D2. 5D3 would still be better of course.

> To just have a matrice multiplication to get from one space into another is right, there is no offset needed? (for having the center (D65) always at the correct position) There are combinations between processing and output gamut, where all becomes red or blue... right.

This is true, colours might shift when adding saturation in the gamuts with D50 white point. It might not be very noticeable, but could be compensated for by using D65 as saturation mid point instead of the processing gamut's default white/grey. (can't explain too well)

> I could compare in Resolve. I have no idea how to do in Lightroom (is it possible here?).

What bouncyball said ^

**Comment by dannephoto** [comment]
What! Not having the eosm as a first cam ;). Too bad your 5D2 got broken. Sad to hear. 
I did a few tests and had a few ideas too. White balance is still problematic. I can´t fix it. Too complicated but in time I think it will be fixed.

Check following pictures downloadable here:
Original
![01_wb_5400 png_800px](https://user-images.githubusercontent.com/22038358/55791485-df690a80-5abe-11e9-95f9-e4b790d57902.jpg)

**Comment by dannephoto** [comment]
What! Not having the eosm as a first cam ;). Too bad your 5D2 got broken. Sad to hear. 
I did a few tests and had a few ideas too. White balance is still problematic. I can´t fix it. Too complicated but in time I think it will be fixed.

Check following pictures downloadable here including MLV and 3D lut:
https://bitbucket.org/Dannephoto/mlv_app_compiler/downloads/test_aces.zip

Original
<img width="1015" alt="01_wb_5400" src="https://user-images.githubusercontent.com/22038358/55791562-10493f80-5abf-11e9-9ef8-a04a1ddc88eb.png">

iso_2300
<img width="1007" alt="02_wb_2300" src="https://user-images.githubusercontent.com/22038358/55791596-1e975b80-5abf-11e9-8aa4-69eb1694b1f9.png">

<img width="1013" alt="bild" src="https://user-images.githubusercontent.com/22038358/55791604-27882d00-5abf-11e9-9403-2d2f7ea3fdea.png">

03_wb_2300_unscience_color_fix
<img width="1013" alt="03_wb_2300_unscience_color_fix" src="https://user-images.githubusercontent.com/22038358/55791618-31119500-5abf-11e9-9e97-5216a8486bfe.png">

04_aces_ap0_rec2020
<img width="1014" alt="04_aces_ap0_rec2020" src="https://user-images.githubusercontent.com/22038358/55791627-38d13980-5abf-11e9-92b4-029892f107f1.png">

05_Aces_ap0_Aces_ap0
<img width="1009" alt="05_Aces_ap0_Aces_ap0" src="https://user-images.githubusercontent.com/22038358/55791642-3ff84780-5abf-11e9-8063-1861e1083756.png">

06_Aces_ap0_Aces_ap0_lut aces_to_rec709
<img width="1005" alt="06_Aces_ap0_Aces_ap0_lut aces_to_rec709" src="https://user-images.githubusercontent.com/22038358/55791653-4686bf00-5abf-11e9-943d-ddd6798c4176.png">


I wonder if we should have "passthrough" options in the color space rolldown window? If we select aces gamut in first we migh wanna keep that exact setting on export and use 3D luts in post. 
Also. If we select ap0 in first roll down and rec2020 in output, colors should be perfect coming back or am I wrong here?

**Comment by dannephoto** [comment]
> I wonder if we should have "passthrough" options in the color space rolldown window? If we select aces gamut in first we migh wanna keep that exact setting on export and use 3D luts in post. 
> Also. If we select ap0 in first roll down and rec2020 in output, colors should be perfect coming back or am I wrong here? 

**EDIT:** Answering myself. Yes rec2020 don´t specify color gamut. But how do we specify color gamut on top of aces if that was selected? Should we have the aces_to_rec709 gamut or aces_to_alexa_wide gamut etc in the second rolldown window?

**Comment by ilia3101** [comment]
Yes there's just pass through if the same colour spaces are selected. I think the wonky colour issues are because I haven't sorted out processing to make it linear at the final conversion stage (you'll notice i do a stupid gamma thing there that's slow and bad)

**Comment by ilia3101** [comment]
@dannephoto thanks for the adobe comparison. Looks smoother, but it is both colder and less saturated. Could you try and match the colours with MLV app a little closer? Using saturation and whitebalance in whatever way required, I just want to see how it rolls of those highlights in the background.

**Comment by ilia3101** [comment]
Off topic, but the MLV App tonemapping causes a bit too much flattening (visible on your comparison with ACR), and I have noticed it makes me grade shots darker than I want often. I will add another reinhard function but affecting a smaller range at the top.

**Comment by dannephoto** [comment]
Aha, I see. Double selection same as passthrough? 
No time atm checking adobe stuff but other than wb breaking in colder kelvin regions mlv app works just fine.

**Comment by ilia3101** [comment]
I must take a complete development break from MLV App until late June for educational purposes. I do want to continue this later of course. It is important. Though if I get a camera (even eos m) earlier than June, I may be partially back before June (can't resist).

Main question:
### Is this branch at risk of becoming unmergeable with the main any time soon?

**Comment by masc4ii** [comment]
Take your time Ilia. If it will be mergeable... I don't know. At the moment it still should be. I do not plan any bigger new features or changes at the moment, so for now it should be safe.

**Comment by ilia3101** [comment]
>  Same for the better blur for Highlights/Shadows. You can switch it in the code only atm. I use it already and I am very satisfied. Maybe it is also okay to just use the better blur... what do you think?

Have you put he better blur in yet as defautl? I think it is a good idea to just have it as a default, no switch. It will only improve all old edited files if anything. I think you should do that and it can be release time. That is enough of a reason to release.

**Comment by ilia3101** [comment]
> I've been thinking of just using rec709 as the processing gamut (after doing tonemapping in a wider one to avoid the ugly clipping). Rec709 sounds limited, but if the lookup tables represented a range of values: something like -1.0 to 3.0, instead of the current 0.0-1.0, the gamut wouldn't be limited to rec709 at all, and would allow final output in wider gamuts. This may increase banding but it's still a really small risk.

I will do this. Just quoting to remember. There will be three selectable gamuts: a initial tonemapping/exposure gamut, processing gamut, rec709 by default (with extended dynamic range like -1 to 3), and a output gamut. If all three are the same it will just pass through with no conversion

**Comment by masc4ii** [comment]
>Have you put he better blur in yet as defautl?

Yes I have. 

>That is enough of a reason to release.

Really? Haha... okay.

**Comment by ilia3101** [comment]
> > Have you put he better blur in yet as defautl?
> 
> Yes I have.

Yay

> > That is enough of a reason to release.
> 
> Really? Haha... okay.

I think, unless you have more stuff to add. I said that because someone on ML forum said something about leaving camera raw after they saw the comparison between old blur and this.

**Comment by masc4ii** [comment]
I prepared a release entry. Have a look.
Some day I would like to have a chromatic aberation correction - but I haven't found a good solution yet... if you watch my latest youtobe vid carefully, you'll see very bad CAs produced by the speedbooster. But I think I won't find a solution until v1.7.

**Comment by dannephoto** [comment]
Speed booze! Can you link?
Off topic:
Trying to compile mlv app for windows. Anybody succeded with mingW on mac?

**Comment by masc4ii** [comment]
@bouncyball-git tried on Linux with success. Win64 static version is build this way, if I understood right.
https://www.youtube.com/watch?v=9cH8ZO-DbUo

**Comment by dannephoto** [comment]
Nice! Sound of music 2.0 :).
Will check later on my lap top.

**Comment by ilia3101** [comment]
Chromatic aberration is an interesting problem

**Comment by bouncyball-git** [comment]
Great VID Markus!

Hmm,,, I did not know there is some sort of speedbooster for canon EF-M? Do you have Viltrox?

**Comment by masc4ii** [comment]
@ilia3101 : I am not sure if the new Resolve beta has it - I think I read something about, but I am not sure if this feature is free.

I found this repos, but haven't tried yet: https://github.com/vicrucann/corrCA-prototype
And this paper looked interesting too, no idea if we get something like that implemented:
https://www.researchgate.net/publication/221159248_Chromatic_aberration_reduction_through_optical_feature_modeling
Maybe a starting point...

@bouncyball-git : thanks 😄  Yes, it is the Viltrox Speedbooster + EOS M + EF 16-35 2.8 II.

**Comment by bouncyball-git** [comment]
> Win64 static version is build this way, if I understood right.

Yes that's right. Compiled with mingw64 under Linux and linked with statically compiled QT/etc libs.

**Comment by dannephoto** [comment]
Trying to get it working under mac. Started out yesterday and now I am waiting for mxe to compile dependencies. Only been 6 hours wait so far. Crap!

**Comment by bouncyball-git** [comment]
Yup MXE is the way to go :). QT compiling it extremely annoying for sure hahaha... that is why I'm using already compiled 5.9.1 all the time (at least it is working w/o probs so far)

**Comment by ilia3101** [comment]
> > Win64 static version is build this way, if I understood right.
> 
> Yes that's right. Compiled with mingw64 under Linux and linked with statically compiled QT/etc libs.

impressive

**Comment by dannephoto** [comment]
Crosscompiled mlv_dump.exe but further than that no way ;).
Remember mlv_dump.exe @bouncyball-git? Those were the days. And also your gamma fixes in dcraw :)

**Comment by bouncyball-git** [comment]
Yeah :grin: ... Of course I remember that! You showed me the right way :)

**Comment by masc4ii** [comment]
> > > Have you put he better blur in yet as defautl?
> > 
> > 
> > Yes I have.
> 
> Yay
> 
> > > That is enough of a reason to release.
> > 
> > 
> > Really? Haha... okay.
> 
> I think, unless you have more stuff to add. I said that because someone on ML forum said something about leaving camera raw after they saw the comparison between old blur and this.

v1.7 is commited. Will upload builds soon. @bouncyball-git : 😉 could you pleeeeeaaaase again build 64bit win version? 😉

**Comment by bouncyball-git** [comment]
Done! :grin: 

32/64bit uploaded

**Comment by masc4ii** [comment]
One question: did you include the new ffmpeg from ffmpeg folder? Not sure if it installs automatically with Qt make button/command.

**Comment by bouncyball-git** [comment]
nope sorry...

**Comment by masc4ii** [comment]
I could download, exchange ffmpeg in zip, upload.

**Comment by masc4ii** [comment]
All is fine now. Thanks!

**Comment by masc4ii** [comment]
> Chromatic aberration is an interesting problem
> 
> I haven't seen many (or even any) open source programs with good solutions.

@ilia3101 : this is a open source solution, which should work fine. It works on RAW data. But no idea so far how to include here.
https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/CA_correct_RT.cc

**Comment by bouncyball-git** [comment]
Do you have Idea how adapt its interface to mlvapp?

**Comment by masc4ii** [comment]
The interface adaption would be easy, we just have to convert array2d<float> to float** (input/output). The bigger problem is to get all the dependencies minimized and compiled. Here I gave up after some hours for now.

**Comment by bouncyball-git** [comment]
Hmm... raw values have to be converted to float, processed and then converted back to int16/14?

**Comment by masc4ii** [comment]
Our RAW values are float when be leaded into debayer. So it could be placed one step before.

**Comment by masc4ii** [comment]
Yay... file from last link with some additional little changes compiles. I am able to call the function and the algorithm does something... it produces CA. Hm. Something might be wrong. But some progress... 😄

**Comment by masc4ii** [comment]
Yay... auto correct doesn't work for now, but manual correction... have a look! And speed is really not bad!
![Bildschirmfoto 2019-05-14 um 12 57 06](https://user-images.githubusercontent.com/30245296/57692804-fcf23c80-7647-11e9-9048-7f04b6d3105c.png)
![Bildschirmfoto 2019-05-14 um 12 56 45](https://user-images.githubusercontent.com/30245296/57692811-0085c380-7648-11e9-9830-15b07cbcf8a4.png)

**Comment by bouncyball-git** [comment]
Fantastic!

**Comment by masc4ii** [comment]
You know what's funny? CA correction needs the same WB conversion as the debayers. :) Maybe we should create 2 functions: WBconvert and WBconvertUndo, which are located outside debayering and correction code. Here another example (not perfect, but better than nothing):
From:
![Bildschirmfoto 2019-05-15 um 12 01 24](https://user-images.githubusercontent.com/30245296/57767369-769e2f00-7709-11e9-99a5-4e5aeba606f2.png)
To:
![Bildschirmfoto 2019-05-15 um 12 01 10](https://user-images.githubusercontent.com/30245296/57767386-81f15a80-7709-11e9-838b-265761472b73.png)

**Comment by masc4ii** [comment]
You like to try it out? I hope that's all for now. https://github.com/ilia3101/MLV-App/commit/7af33b72390445dac656ee3e884df12911475cc7

**Comment by bouncyball-git** [comment]
I think the result is very good!

Regarding wbaply/wbundo... I think we should find the way to do WB once and do not undo it anywhere.

**Comment by bouncyball-git** [comment]
Commited small patch to use stdint.h and correctly compile under Linux.

Maaan! So many warnings added because of some variables not used...

**Comment by masc4ii** [comment]
Thank you @bouncyball-git ! Haven't tried on Linux or Windows yet. 
I think it is not difficult to get rid of all the warnings. Crazy, yes 😄  
When applying correct WB only once before debayer (therefor not in processing), the result looks not good. So I think we have to undo once like in the past. Or I did something wrong, hehe. Got best results at 6500K WB for debayer.

**Comment by bouncyball-git** [comment]
> When applying correct WB only once before debayer (therefor not in processing), the result looks not good.

When WB process shifts values of some channel(s) they prone to go out of the bounds of uint16 after converting post debayer. AFAIK that was the main reason why @ilia3101 restored original vals.

There should be the way to have those values in bounds without using floating point during entire processing pipeline... OR we should try to use experimental all floating point pipeline and evaluate the speed difference. Once @ilia3101 said he wanted to try this out...

**Comment by ilia3101** [comment]
> Regarding wb_apply/wb_undo... I think we should find the way to do WB once and do not undo it anywhere.

Doing white balance on the bayer data is (allegedly) less accurate than what we have now- doing it in the LMS colour gamut after debayering (LMS, where each channel corresponds to an eye cone- long medium short, Adobe does this)

I think RawTherapeee might be doing white balance on bayer channels, but I would rather copy Adobe's method's than rawtherapee's.

So I do think we should have an undo after debayer.

> There should be the way to have those values in bounds without using floating point during entire processing pipeline... OR we should try to use experimental all floating point pipeline and evaluate the speed difference. Once @ilia3101 said he wanted to try this out...

Yes I did and still do. I've also really been wanting to re write the processing with OpenGL. I know, everyone goes on about CUDA and OpenCL, but I've realised OpenGL fragment shaders are perfect for this kind of processing, will also be efficient for outputting directly to the display.

**Comment by bouncyball-git** [comment]
> ... but I would rather copy Adobe's method's than rawtherapee's.
> So I do think we should have an undo after debayer.

Thanks for clarification!

> Yes I did and still do. I've also really been wanting to re write the processing with OpenGL. I know, everyone goes on about CUDA and OpenCL, but I've realised OpenGL fragment shaders are perfect for this kind of processing, will also be efficient for outputting directly to the display.

Yes once I tried to use some basic shaders of mlrawviewer with QT opengl stuff (it was fun and was long time ago)... this is doable and will work for sure. However more complex processing than mlrawviewer's gonna be lot more difficult to realize.

**Comment by masc4ii** [comment]
With this commit https://github.com/ilia3101/MLV-App/commit/88e803d1033640953edd284de4bccdae88645c63 I moved the WB convertion to another .c file, so it is identical for all debayers now. Debayers are cleaned from the code. Bilinear, easy and "none" debayer don't need it (just costs time).
@ilia3101 : I also integrated blacklevel correction. Have a look.

**Comment by ilia3101** [comment]
> > ... but I would rather copy Adobe's method's than rawtherapee's.
> > So I do think we should have an undo after debayer.
> 
> Thanks for clarification!
> 
> > Yes I did and still do. I've also really been wanting to re write the processing with OpenGL. I know, everyone goes on about CUDA and OpenCL, but I've realised OpenGL fragment shaders are perfect for this kind of processing, will also be efficient for outputting directly to the display.
> 
> Yes once I tried to use some basic shaders of mlrawviewer with QT opengl stuff (it was fun and was long time ago)... this is doable and will work for sure. However more complex processing than mlrawviewer's gonna be lot more difficult to realize.
> 
> Edit: and the reason why high quality image processing applications use OpeCL/CUDA instead of OpenGL - IS: that OpenGL implementations have some "corner cutting" for speed's sake, like not doing math in double precision etc.

OpenGL does have "corner cutting" , worse precision, but I still think it would be more precise than the 16 bit look up tables right now. You can also set highp mediump lowp in fragment shader, don't know how much improvement that does.

It will be difficult to do filters like the RBF , but not too bad for most of the processing, some of it like curves could also be made in to a look up table for speed.

Also do the master and BetterProcessing branches still merge without much difficulty ?

**Comment by masc4ii** [comment]
>Also do the master and BetterProcessing branches still merge without much difficulty ?

Hard to say... some time is gone now. But I think it should somehow be mergeable with some work. I tried to not change parts in Master I changed in BetterProcessing. So... hmmm... but sometimes GitHub is strange ;-)

**Comment by ilia3101** [comment]
Ok I have a bit of an idea, could we merge it and disable options in the interface? For now just have it set to sRGB and sRGB to keep things looking the same.

**Comment by masc4ii** [comment]
Hm... but is this the same? When trying it looks somehow different. 
However... ready for v1.8 before? ^^

**Comment by ilia3101** [comment]
I seem to remember one combination did produce idential to old results, but I'll have to check it out. Yeah I think 1.8 would be good to release about now.

**Comment by ilia3101** [comment]
1.8?

**Comment by masc4ii** [comment]
https://github.com/ilia3101/MLV-App/commit/209559feb5305e375db5c72ff4c083da66c683ca

**Comment by masc4ii** [comment]
@bouncyball-git : Could you again please build the Win64 version? Don't forget the newer ffmpeg (same as in v1.7) and all the focus pixel maps (from the folder in src/...). I now can also build a Win64 dynamic version with Qt5.12, but Qt has some very odd bugs in this version, so I don't like to upload that.

**Comment by bouncyball-git** [comment]
I'll do later today.

**Comment by bouncyball-git** [comment]
Done :)

**Comment by masc4ii** [comment]
Thank you! :)

**Comment by ilia3101** [comment]
I am feeling a little bit unhappy about how the BetterProcessing branch has turned out right now, there is a really unpleasant stage at the end where it does a final colour space conversion, but the image is not linear at all by that stage, and this whole time I have been dreading figuring out how to get it back to linear for that conversion... it would require untangling many of the stages leading up to it, difficult as I don't remember how much of it works (don't know what the gamma or anything is by then).

I was even planning to add a third stage with another separate gamut option (but now I realise this is just too much).

I think what I was doing was much more complicated than it needed to be.

I'm thinking much simpler now. What I would like to do is: go back to master for this, and simply add an option for processing gamut and tonemapping function (only TWO extra options, to replace profile), nothing else. No extra colour conversion at the end. This will leave processing code cleaner, and easier to remaster later on, which I REALLY want to start doing, processing has outgrown the original version with it's stupid if conditions for lookup tables that switch what they do based on exposure and all that.

@masc4ii I am not sure what to do right now. I know that you put in a lot of effort on the GUI part already, so it may be annoying, but I really want to start again. I am sorry about this.

Hopefully this will be much much easier to implement in the GUI.

**Comment by ilia3101** [comment]
Changes that will be required from Qt side:

GUI:
- Delete profile combobox
- Add curve combobox, which will set tonemapping functions and LOG profiles
- Add gamut selector combobox, will select gamut (will copy the constants from BetterProcessing)
- When selecting LOG curves, automatically set corresponding gamut, and disable creative adjustments (you can tell me if you think this is a bad idea)
- For now, camera matrix stuff will stay as is (I think)

MASXML:
- Processing profile option should get replaced with gamut and curve
- Old masxml files with profile should be converted like this: set curve based on profile, gamut to rec709, and keep camera matrix option as is. Will stay identical this way for those files.

**Comment by ilia3101** [comment]
Edit: Ok there will also need to be a gamma slider in the GUI and a value in MASXML. For old MASXML files with profile, it will need to be set to what the profile said in [image_profiles.c](https://github.com/ilia3101/MLV-App/blob/9b3d31613ced492a79d3ca8faf130c825077acbf/src/processing/image_profiles.c)

**Comment by ilia3101** [comment]
Why I said all that: I want to make a smaller step. Doing that will allow easier refactoring, which will allow for even bigger improvements.

**Comment by dannephoto** [comment]
Great thinking Ilia3101. Every now and then a full restart will be the better way. Hell I refactored cr2hdr-r top MLP and finally it bacame Switch. Every time a full revision was needed. Boy did I learn...

**Comment by masc4ii** [comment]
Okay, let's do it, @ilia3101 . But I would recommend to open again a new branch for that. If we have luck this time, we merge to master branch!

No problem with the GUI part. We'll get that ;-) If I remember right, there were some strange things in my implementation, so maybe it is better to not use this.

**Comment by ilia3101** [comment]
I will soon open the new branch soon.

**Comment by ilia3101** [comment]
Will use ImprovedProcessinf branch, can't delete branches so typo name will have to do.

**Comment by ilia3101** [comment]
@masc4ii Do you use uncolourscience fix often?

**Comment by ilia3101** [comment]
One thing I'd like to do is remove the ban on unselecting "allow creative adjustments" in normal non log profiles.

**Comment by ilia3101** [comment]
After the changes i push soon... Qt side should be modified to check after setting profile if the processing allows creative adjustments, instead of controlling that itself.

**Comment by ilia3101** [comment]
clarity slider should be greyed out when disable creative adjustments, I will add the internal disabling myself

**Comment by ilia3101** [comment]
AS well as highlights and shadows

**Comment by ilia3101** [comment]
Ok I have commited some improvements, just disable creative adjustments needs some fixing up on UI side (I think). I'm definitely in my zone.

**Comment by ilia3101** [comment]
Don't convert profiles in to gamut and tonemapping function just yet. Keep profiles there at first, they are already better on the inside.

**Comment by ilia3101** [comment]
If anyone wants, you can test things out right now, have a look at the improved log profiles with correct gamuts! Now the colours will actually match an Aalexa in Log C...

**Comment by masc4ii** [comment]
Puh... so much input :-D 
Will see what I can do... this week is very full, but I'll try to support!

**Comment by ilia3101** [comment]
Ok thanks! Don't worry about reading everything I wrote

Right now the main issue is "allow creative adjustments" option seems to be not working. I think I remember this issue coming up in the original BetterProcessing.

**Comment by masc4ii** [comment]
Have a look please... tried to fix it. Problem: loading old projects will disable "Allow creative adjustments", because it was not force enabled in files... so you'll have to manually enable it now. For newly imported MLVs, it is force enabled.

**Comment by masc4ii** [comment]
> Changes that will be required from Qt side:
> 
> GUI:
> 
>     * Delete profile combobox
> 
>     * Add curve combobox, which will set tonemapping functions and LOG profiles
> 
>     * Add gamut selector combobox, will select gamut (will copy the constants from BetterProcessing)
> 
>     * When selecting LOG curves, automatically set corresponding gamut, and disable creative adjustments (you can tell me if you think this is a bad idea)
> 
>     * For now, camera matrix stuff will stay as is (I think)
> 
> 
> MASXML:
> 
>     * Processing profile option should get replaced with gamut and curve
> 
>     * Old masxml files with profile should be converted like this: set curve based on profile, gamut to rec709, and keep camera matrix option as is. Colours will stay identical this way.
> 
> 
> We will not do this all at once, but in a couple of smaller steps.

All this has still to be done... Until now, it works (looks in GUI mostly) like master branch but with your new code.

**Comment by ilia3101** [comment]
Thanks for the fix!

> Until now, it works (looks in GUI mostly) like master branch but with your new code.

Looks like master branch, but a good first step. Just as I wanted. If you are very busy, maybe I will see how much I can do with Qt, maybe not much though.

**Comment by masc4ii** [comment]
In the evenings I will find a little time... If I remember right, this combobox rules were a little difficult. From Qt side it is easy, but my own concept of how the elements work together was a little more complex.

**Comment by ilia3101** [comment]
> In the evenings I will find a little time... If I remember right, this combobox rules were a little difficult. From Qt side it is easy, but my own concept of how the elements work together was a little more complex.

It hope it will be easier this time, as there doesn't need to be a main profile selector that also says 'custom' if the user makes a change. Their just needs to be curve, gamut and gamma controls, which should call `processingSetTonemappingFunction`, `processingSetGamut` and `processingSetGamma` respectively.

Only extra thing: when selecting a log curve, automatically set it's corresponding gamut, gamma to 1.0, and disable creative adjustments - but don't stop the user from changing any of those settings afterwards. You can look at image_profiles.c for reference. This should only happen for log profiles.

**Comment by masc4ii** [comment]
Please test, @ilia3101 . I don't see any difference for gamut combobox. The other elements seem to work. Receipt and session is not supported yet... same for export. You just can play around in GUI for now.

**Comment by ilia3101** [comment]
It is working!!! I will fix gamut issue in a minute.

**Comment by masc4ii** [comment]
Yay... now it works. Thanks. But somehow, I just see a difference in gamut for clips recorded in tungsten ~2500K. Is that normal?

**Comment by ilia3101** [comment]
Hmm could you show what you mean. Sorry don't have a tungsten clip on me right now.

**Comment by ilia3101** [comment]
Feeling slightly devastated... just found a clipping bug that has been around since camera matrix has been used. It can be seen in Linear mode, it does not clip to white :( not sure where it comes from. Good opportunity to try out float processing however :)

**Comment by ilia3101** [comment]
I am starting to think that float may even be faster than lookup tables in most cases (I realised how slow memory is), especially if we can make the compiler vectorise it. Maybe float processing should be a new issue.

**Comment by masc4ii** [comment]
Maybe I post a little testrun this evening. For clips in daylight, I don't see any difference between all gamuts. In tungsten with rec.709 it looks the same like in master branch (highlights clip to cyan), aces makes it a little better, best seems to be bmdfilm, while with canon-c-log I get a partially inverted picture.

**Comment by ilia3101** [comment]
> Maybe I post a little testrun this evening. For clips in daylight, I don't see any difference between all gamuts. In tungsten with rec.709 it looks the same like in master branch (highlights clip to cyan)

Glad to hear it gets the same results and nothing is broken.

> aces makes it a little better, best seems to be bmdfilm, while with canon-c-log I get a partially inverted picture. But colors look different for each gamut in tungsten clips.

Oh yeah, I was going to mention that I think we should exclude the broken canon log function. I expected colours to look different for each gamut.

**Comment by masc4ii** [comment]
Here a little testrun. Left: daylight, Right: ~2500K. Rows: some of the gamuts, left = right.
![Testrun](https://user-images.githubusercontent.com/30245296/63539982-82c13600-c51b-11e9-8dc0-5e5fa84a165a.png)

**Comment by masc4ii** [comment]
Oh... I now see why my test result is as it is .... it does not depend on light! It depends on camera matrix enabled or disabled! Left: disabled, Right: enabled. If I enable it for the left clip, colors look different too!

**Comment by masc4ii** [comment]
Receipt support is implemented. So now you can also copy/paste, change current clip without loosing settings and export.

**Comment by bouncyball-git** [comment]
Guys this is fantastic step forward!!! :grin:

**Comment by ilia3101** [comment]
Is MASXML now saving only Gamut, Tonemapping function and Gamma instead of profile?

**Comment by masc4ii** [comment]
It saves all settings.

**Comment by ilia3101** [comment]
Ok. Profile will interfere with these three new settings, depending if they are applied before or after the profile, so 'Profile' should be removed from interface and MASXML completely, as the 3 new options replace it.

For dealing with old MASXML that do have profile, you could call the profile setting function, or just pick the correct settings for new 3 parameters with some switch block, whichever method you prefer.

To sort of replace profile: when selecting some tonemap functions, the gamut should automatically be set:
Alexa Log C : Alexa Wide Gamut
Cineon Log: Alexa Wide Gamut
Sony S Log: Sony S Gamut 3
BMDFilm: BMDFilm(gamut)
Also disable creative adjustments for those...
But don't stop the user from changing anything back. This would just be a helpful feature for people wanting an accurate log image.

1 more small interface change: camera matrix combobox should be moved in to the same area as these new options, as it is so related.

Sorry I'm typing this on a phone so it may be difficult to read. I hope these suggestions aren't too complicated, tell me if they are or if you disagree with anything.

**Comment by masc4ii** [comment]
> Ok. Profile will interfere with these three new settings, depending if they are applied before or after the profile, so 'Profile' should be removed from interface and MASXML completely, as the 3 new options replace it.

No, no problem. I did it in a way, that you'll get exactly what you saved. (Hopefully. At least I think this is what I implemented.)

> For dealing with old MASXML that do have profile, you could call the profile setting function, or just pick the correct settings for new 3 parameters with some switch block, whichever method you prefer.

I think we should increase MASXML version number and create a new kind of compatibility logic. But I have to do some research for correct compatibilty... for now I still have some problems with "Allow creative adjustments", which was invisibly enabled when no log profile was chosen.
 
> To sort of replace profile: when selecting some tonemap functions, the gamut should automatically be set:
> Alexa Log C : Alexa Wide Gamut
> Cineon Log: Alexa Wide Gamut
> Sony S Log: Sony S Gamut 3
> BMDFilm: BMDFilm(gamut)
> Also disable creative adjustments for those...
> But don't stop the user from changing anything back. This would just be a helpful feature for people wanting an accurate log image.

Sounds interesting. I could try that out. But all those logic increase complexity a lot, because MASXML loading and user input call the same functions. We should be careful to not create problems... but I'll try that!

> 1 more small interface change: camera matrix combobox should be moved in to the same area as these new options, as it is so related.

Oh yes. I think so to... currently I scroll a lot 😄

**Comment by ilia3101** [comment]
> I think we should increase MASXML version number and create a new kind of compatibility logic. But I have to do some research for correct compatibilty

I agree we should increase version number for this change. The compatibility logic shouldn't be too difficult. There is two ways to go about this logic for log profiles: 1 - set correct log/tonemap curve for that profile, and rec709 gamut to match old images exactly, or 2 - set correct log curve *and* correct gamut, this will not match old results exactly, but the intended log profile will be recreated better, assuming that's what people want. Maybe we should just go for 1, to make sure everything stays the same.

> for now I still have some problems with "Allow creative adjustments", which was invisibly enabled when no log profile was chosen.

Not sure what this is, but hopefully it won't be difficult to fix. Is there a chance this problem is happening inside processing or is it on Qt side?

> Sounds interesting. I could try that out. But all those logic increase complexity a lot, because MASXML loading and user input call the same functions. We should be careful to not create problems... but I'll try that!

Ah yes true, those functions should only do that when the user makes a selection, not when loading MASXML.

**Comment by ilia3101** [comment]
MASXML logic for converting from profile to new settings:

| Profile | Tonemapping function | Gamut | Gamma |
|-|-|-|-|
| PROFILE_STANDARD | TONEMAP_None | GAMUT_Rec709 | 3.15 |
| PROFILE_TONEMAPPED | TONEMAP_Reinhard | GAMUT_Rec709 | 3.15 |
| PROFILE_FILM |TONEMAP_Tangent | GAMUT_Rec709 | 3.46 |
|PROFILE_ALEXA_LOG|TONEMAP_AlexaLogC|GAMUT_AlexaWideGamutRGB|1.0 |
|PROFILE_CINEON_LOG|TONEMAP_CineonLog|GAMUT_AlexaWideGamutRGB|1.0|
|PROFILE_SONY_LOG_3|TONEMAP_SonySLog|GAMUT_SonySGamut3|1.0|
|PROFILE_LINEAR|TONEMAP_None|GAMUT_Rec709|1.0|
|PROFILE_SRGB|TONEMAP_sRGB|GAMUT_Rec709|1.0|
|PROFILE_REC709|TONEMAP_Rec709|GAMUT_Rec709|1.0|
|PROFILE_BMDFILM|TONEMAP_BMDFilm|GAMUT_BmdFilm|1.0|

**Comment by DeafEyeJedi** [comment]
Uncoloursciene fix evert single time for the most part. @Danne has nailed this now down nicely!
    On Saturday, August 17, 2019, 2:01:35 PM PDT, Ilia Sibiryakov <notifications@github.com> wrote:  
 
 
@masc4ii Do you use uncolourscience fix often?

**Comment by dannephoto** [comment]
Hi deafeye. So it's here you are :).
Did yiu try out the latest eosm build by the way?

**Comment by masc4ii** [comment]
>@masc4ii Do you use uncolourscience fix often?

Yes I do. For all my tungsten clips 😄 

@ilia3101 : I added the compatibility layer now, and moved the UI elements. When playing with that new version now, I am not sure if the place of the profile groupbox is nice. Should it be where it is, or better below "Processing" (more in the near where some of the elements are in master branch)?

**Comment by ilia3101** [comment]
I think below processing would be better actually, as it will be closer to where the original profile and camera matrix options were.

**Comment by masc4ii** [comment]
Done.

And I've hidden Canon C Log. Maybe we get it corrected one day - no problem to unhide/show it again. 😄

**Comment by ilia3101** [comment]
I removed Canon C log and changed all the numbers. Have you taken that in to account? Just realised that was slightly dumb

**Comment by masc4ii** [comment]
Ups... no, haven't seen. Then I should revert and change the numbers too.

**Comment by ilia3101** [comment]
sorry

**Comment by masc4ii** [comment]
Done. No, you mustn't excuse... it was me not updating latest revision. 😄

**Comment by masc4ii** [comment]
But something seems to be wrong now. Alexa is way too bright...

**Comment by ilia3101** [comment]
Hmm thats really strange, just had a look. Must be an issue inside inside processing. So confusing>

**Comment by masc4ii** [comment]
Looks like canon-c-log is loaded if alexa is chosen.

**Comment by masc4ii** [comment]
static void * tonemap_functions[] =
{
    (void *)&NoTonemap,
    (void *)&ReinhardTonemap,
    (void *)&TangentTonemap,
    (void *)&CanonCLogTonemap,
    (void *)&AlexaLogCTonemap,
    (void *)&CineonLogTonemap,
    (void *)&SonySLogTonemap,
    (void *)&sRGBTransferFunction,
    (void *)&Rec709TransferFunction,
    (void *)&HLG_TransferFunction,
    (void *)&BmdFilmTonemap
}; 


?!

**Comment by masc4ii** [comment]
Yepp... that's it! I commit! ;)

**Comment by masc4ii** [comment]
@ilia3101 : how far are we from what you wanted to have in this branch? If all is done... we like to merge? Or is it still dangerous?

**Comment by ilia3101** [comment]
Ah yes we are very close to being able to merge I think

**Comment by ilia3101** [comment]
I just wrote out a thing about changes that should be made with a whole table but I decided we don't need to over complicate on the Qt side.

Also why is profile still in the MASXML, is it just to set what the combobox will show?

Only change I think should be made now is to have the profile combobox show say like 'custom' if the user makes any changes *after* selecting a profile from there. And if they then choose a profile again, the combobox will show the name of selected profile again, until any more changes are made.

I think "Processing Profile" box should now be be renamed to "Profile Preset"

**Comment by masc4ii** [comment]
>Also why is profile still in the MASXML, is it just to set what the combobox will show?

Exactly. It is just there, that you'll load what you saved.

>I think "Processing Profile" box should now be be renamed to "Profile Preset"

No problem. I'll do.

>Only change I think should be made now is to have the profile combobox show something like 'custom' if the user makes any changes after selecting a profile from there.

Hm. This is what we had in this old branch and it is more complicated than one might expect. "Custom" is then always available and can be chosen. This is the only setting, where nothing should happen to other controls - easy so far. But as soon as you change another setting (also gamma) the combobox must automatically switch to custom. Choosing a preset profile changes the parameters to their default value, and here this parameter actions mustn't control "Custom" backwards (but it does by standard, we can forbid under cases). This can all be implemented... but it isn't just a few lines. I still remember this fun... 😄

**Comment by ilia3101** [comment]
> > Also why is profile still in the MASXML, is it just to set what the combobox will show?
> 
> Exactly. It is just there, that you'll load what you saved.
> 
> > I think "Processing Profile" box should now be be renamed to "Profile Preset"
> 
> No problem. I'll do.
> 
> > Only change I think should be made now is to have the profile combobox show something like 'custom' if the user makes any changes after selecting a profile from there.
> 
> Hm. This is what we had in this old branch and it is more complicated than one might expect. "Custom" is then always available and can be chosen. This is the only setting, where nothing should happen to other controls - easy so far. But as soon as you change another setting (also gamma) the combobox must automatically switch to custom. Choosing a preset profile changes the parameters to their default value, and here this parameter actions mustn't control "Custom" backwards (but it does by standard, we can forbid under cases). This can all be implemented... but it isn't just a few lines. I still remember this fun...

Wow I caught this just 28 seconds later.

When opening a MASXML, the "profile" does not need to be applied, as the tonemap function, gamut, and gamma do what profile does. The only use of profile right now is to act as a preset for those three settings. It won't do any harm if you call processingSetImageProfile() before setting the three settings.

idea:
Would it help if I add a PROFILE_CUSTOM as -1? - that doesn't actually make any changes. And whenever the user changes gamma/tonemap/gamut, just set profile to custom and it won't do any harm.

**Comment by ilia3101** [comment]
Actually I can imagine the difficulties you might have even this way :/

**Comment by masc4ii** [comment]
Until here I had no problem at all... easy going. But I'll try "Custom".

**Comment by ilia3101** [comment]
I imagine the problems come from setting parameters through UI controls that affect others, and it just becomes a recursive kind of issue (am I right?). I think I remember this feeling from various times I have done UIs. If this is too painful we could just keep it the way it is, it's not too bad I think.

The reason I wanted this 'custom' profile was to make it clear to users that once they change settings after setting a profile, it is no longer that profile.

**Comment by masc4ii** [comment]
>and it just becomes a recursive kind of issue (am I right?)

Correct... the controls control themselfs. Currently, when selecting a preset, you'll get "Custom" a millisecond later, haha.

**Comment by masc4ii** [comment]
With file loading we will always have "Custom", however you saved. MASXML loading will set all parameters. So this will set preset to Custom.

**Comment by masc4ii** [comment]
Woah... no idea how I solved this last time. I think I re-implemented all presets again in GUI part, to make those stupid UI elements more intelligent. Works as long as no preset changes or as long there is no new preset added / deleted...

**Comment by ilia3101** [comment]
> With file loading we will always have "Custom", however you saved. MASXML loading will set all parameters. So this will set preset to Custom.

Hmm this was one of the possibilities I imagined might happen. Sounds alright to me, better than what we had before.

> Woah... no idea how I solved this last time. I think I re-implemented all presets again in GUI part, to make those stupid UI elements more intelligent. Works as long as no preset changes or as long there is no new preset added / deleted...

Ahh, it would actually be good to have the presets in UI - as they are not really a meaningful thing in processing anymore, they're just a bit of help for the user. I would then remove the presets from processing if you did it this way.

Is this something that would take a lot of time though? I think it would be good as a next step in MLV App, maybe even have user be able to make presets.

We don't have to do this now, could keep it as it is and make open it as an issue for later.

**Comment by masc4ii** [comment]
Please let it in the processing part for now. This works not so bad atm. I gave up for now with this "Custom" solution. I have another idea, let me please know what you think about. The UI looks like this:
![Bildschirmfoto 2019-09-06 um 21 00 43](https://user-images.githubusercontent.com/30245296/64453783-a672a780-d0e9-11e9-97bd-59e904bc2617.png)
You click on "Select profile..." and you'll see all profiles. After you've chosen, this element goes back to "Select profile...", but all other elements are set. Pro: we don't have to automate any element more than we have now. Con: MASXML loading must set this preset element for version <= 2, and has not to set it for version >2. So... seems solveable!?

**Comment by masc4ii** [comment]
Oh no... other problems with that... :(

**Comment by ilia3101** [comment]
Ahh missed this

**Comment by ilia3101** [comment]
I quite like the idea of a "select preset" type of menu, if it just says that all the time. Then we don't even need to save the profile at all.

**Comment by ilia3101** [comment]
I definitely like the select preset idea. This thing should just call profile set function, then update the controls in the profile area that have changed. This way we do not have to save 'profile' field at all, as it will not be displayed anyway.

**Comment by ilia3101** [comment]
is this the way you did it?

**Comment by masc4ii** [comment]
Yep, this is what I did. Works not bad for all new sessions. 
The problem is compatibility to MASXML v2. This saves just the profile. This information only is converted if clip is currently loaded into viewer. So, if you load a v2 session with n clips, don't open another clip, but save it into v3 MASXML, you'll get one correct clip (the one which was visible in viewer before) and n-1 wrong clips (in terms of profile).

**Comment by ilia3101** [comment]
maybe MASXML version updating should be done completely at the start? like when loading a MASXML, check the version, and if it is V2 or older, update each one of the clips to latest version, and only after that, actually load the whole MASXML.

**Comment by masc4ii** [comment]
In the end this is nearly what I do. It loads the file into the internal structure, like the old version did. Then when selecting the clip, the processing logic is used and now it is v3-like. And as long you don't save and go through all clips, all seems to be fine. The problem is, if you forget watching a single clip in your session after loading a v2, save to v3 and reopen, this clip will be wrong.
But I can't completely convert on load... difficult thing to explain.

**Comment by masc4ii** [comment]
Yay... if you don't know how to realize: go to bed and try another day. I think it works now. Could you test?

**Comment by ilia3101** [comment]
sorry for my late response! I will test right now

**Comment by ilia3101** [comment]
It seemed to work right with an old MASXML , it was all originally tonemapped or film, didn't test with LOG files though. But I think it's working good.

**Comment by ilia3101** [comment]
If you are think it works well enough I think we can merge any time now.

**Comment by ilia3101** [comment]
Do you think it's safe and ok to merge now?

**Comment by ilia3101** [comment]
https://github.com/ilia3101/MLV-App/pull/176

**Comment by ilia3101** [comment]
I noticed a small issue, when selecting for example Alexa Log preset (or any profile that doesn't allow creative adjustments by default), it does not automatically grey out the controls, even though they are disabled and the creative adjustments checkbox is turned off.

**Comment by masc4ii** [comment]
Oh really... I'll have a look. Thanks for testing! After I fixed, we could merge, yes. :)

**Comment by masc4ii** [comment]
Good... I think it was simple. Forgot to delete one stupid variable from old version.

**Comment by ilia3101** [comment]
I merged it. I hope it was ready :)

**Comment by masc4ii** [comment]
Yes, it was. And it seems still to work!

**Comment by masc4ii** [comment]
Will next version be 2.0, or 1.9? I prepared some points in the release tab... edit whatever you like (did not know how to describe this thing here correctly).

**Comment by ilia3101** [comment]
Let's keep it 1.9 for now. I will have a look at the release. I am ready for release any time now I think.

**Comment by masc4ii** [comment]
What exactly was/is the reason, that gamut setting changes colors just if CameraMatrix is enabled?

**Comment by ilia3101** [comment]
oh yeah about that... maybe greying out gamut selector when not using camera matrix is a good idea.

And the reason for this behaviour is: selecting gamut can work correctly because we know the camera's own gamut, from the camera matrix, therefore converting it to another gamut can be done right. However if we are not using camera matrix, we have no knowledge of the camera's gamut, so converting the camera's colour in to other gamuts makes no sense (because we "don't know" what the initial gamut is).

I could have decided that camera colour, when camera matrix is disabled, will be always interpreted as being in a certain gamut already, but I didn't want to do that, as all cameras are different.

**Comment by masc4ii** [comment]
Ah great. Understood. Will write that to the Wiki. I think this question will be asked many times. Greying out gamut setting is a good idea.

**Comment by masc4ii** [comment]
Done.
Gamut is now greyed out if "Dont Use Matrix".
Wiki: https://github.com/ilia3101/MLV-App/wiki#44-profile

**Comment by masc4ii** [comment]
Okay... I think this (longest?) thread can now be closed... seems to work fine?! :) If not, feel free to reopen!

**Comment by ilia3101** [comment]
good night
